server/datastore/mysql/software.go

```diff
@@ -1307,55 +1311,74 @@ func (ds *Datastore) SyncHostsSoftware(ctx context.Context, updatedAt time.Time)
 		return ctxerr.Wrap(ctx, err, "reset all software_host_counts to 0")
 	}
 
-	// next get a cursor for the global and team counts for each software
-	stmtLabel := []string{"global", "team"}
-	for i, countStmt := range []string{globalCountsStmt, teamCountsStmt} {
-		rows, err := ds.reader(ctx).QueryContext(ctx, countStmt)
-		if err != nil {
-			return ctxerr.Wrapf(ctx, err, "read %s counts from host_software", stmtLabel[i])
-		}
-		defer rows.Close()
-
-		// use a loop to iterate to prevent loading all in one go in memory, as it
-		// could get pretty big at >100K hosts with 1000+ software each. Use a write
-		// batch to prevent making too many single-row inserts.
-		const batchSize = 100
-		var batchCount int
-		args := make([]interface{}, 0, batchSize*4)
-		for rows.Next() {
-			var (
-				count  int
-				teamID uint
-				sid    uint
-			)
+	db := ds.reader(ctx)
 
-			if err := rows.Scan(&count, &teamID, &sid); err != nil {
-				return ctxerr.Wrapf(ctx, err, "scan %s row into variables", stmtLabel[i])
+	// Figure out how many software items we need to count.
+	type minMaxIDs struct {
+		Min uint64 `db:"min"`
+		Max uint64 `db:"max"`
+	}
+	minMax := minMaxIDs{}
+	err := sqlx.GetContext(
+		ctx, db, &minMax, "SELECT COALESCE(MIN(software_id),1) as min, COALESCE(MAX(software_id),0) as max FROM host_software",
+	)
+	if err != nil {
+		return ctxerr.Wrap(ctx, err, "get min/max software_id")
+	}
+
+	for minSoftwareID, maxSoftwareID := minMax.Min-1, countHostSoftwareBatchSize; minSoftwareID < minMax.Max; minSoftwareID, maxSoftwareID = maxSoftwareID, maxSoftwareID+countHostSoftwareBatchSize {
```

Comment:
Shouldn't this be:
`for minSoftwareID, maxSoftwareID := minMax.Min-1, minMax.Min-1+countHostSoftwareBatchSize?`
?